### ISSUE TITLE
Events dropping when it arrives at Digest agent, more specifically, agent memory

### ISSUE THREAD

**Issue by Kristovich** [huginn/huginn#2389]
So I've been battling this off and on for a few months now. 

**My setup**
I'm running huginn/huginn-single-process in docker with 11 worker containers. 

**What I'm trying to do:**
Send about 20 events to digest agent to aggregate into an array. The events I want to aggregate have 2 keys each; values being a title and a tinyurl (these are youtube videos). So there isn't a lot of data to store into memory.

**The problem**
Digest agent ends up sending out around a random number of events from 12-19. It seems to drop the rest.

**Troubleshooting**
At first I thought my events were not fully getting through to the Digest agent. But I checked event output for all my agents and that is not the case. 'Propagate  Immediately' does not affect the result.  I finally created a JavaScript Agent that basically does the same thing and it suffers the same exact problem. I linked both my Javascript digest and the official Digest agent into the same source and they will both drop about 6 events but the dropped events would be different. So, that rules out any possible character formatting that is disliked by memory. This leads me to believe that something is wrong with agent memory when receiving 12+ events quickly. 

Any help would be appreciated. To replicate all you need is a Digest agent receiving/aggregating about 20 events to send out in 1 big digest. 
Also, here is the code to my JavaScript digest agent:
```
Agent.check = function() {
    var digest = []; 
    var eventIndex = this.memory('eventIndex') || 0;
    for(var i = eventIndex -1; i != -1; i--){
     digest.push(JSON.parse(this.memory(i)));
    
    }
    if (digest.length != 0){
    this.createEvent({'events': digest});
    }
};

Agent.receive = function() {
  var events = this.incomingEvents();
  var eventIndex = this.memory('eventIndex') || 0;
  this.memory(eventIndex,JSON.stringify(events[0].payload));
  this.memory('eventIndex', eventIndex + 1);

}
```

**Comment by dsander** [comment]
Hi @Kristovich,

thanks for the detailed explanation of your problem! 

> This leads me to believe that something is wrong with agent memory when receiving 12+ events quickly.

I think you are right, the fact that you said **quickly** and you are running a lot of workers could explain the problem. We don't have any kind of uniqueness for the background jobs. You said that propagate immediately does not improve the situation, is the DigestAgent receiving events from multiple sources?

I think the DigestAgent is receiving multiple Events at the same time, and when it runs in multiple workers at the same time each running `receive` function overrides the changes to the memory of the other running Agents.

**Comment by Kristovich** [comment]
Thanks for the reply, Digest agent is receiving from 1 source. Memory being overwritten does sound like the most probable cause. Is there any way to correct this? Perhaps locking an agent to 1 worker or something similar?

**Comment by dsander** [comment]
I have been thinking about how we can prevent that, if we would not be using `delayed_job` we could use a background job uniqueness logic to ensure only one `AgentReceiveJob` per agent is running at the same time.

The only idea I have at the moment is to add some custom database level lock in the `AgentReceiveJob`, which also isn't perfect because the first job that is started for the Agent will delay every other job and remove time from it's max runtime.

@knu Do you have other ideas how we can prevent multiple Agents from running at the same time?

**Comment by Kristovich** [comment]
So I updated my javascript digest script to drop less events by locking memory. I had to move stored events into credentials because the agents would erase things when setting the memory lock. It's definitely hacky (especially since I had to input random sleep times for workers) but still works slightly better than before. 

Also, I think i found another bug in the process. Originally, I was going to set the memory lock in credentials but it seems like agents cache credential key-value pairs in memory because I could not read an updated credential after its been read once before. I wanted to create a memory 'stack' but I couldn't figure out a way to push an update into credentials(or memory) without copy-update-paste the entire thing. 

Anway, here's the code for anyone experiencing similar issues. Instead of 6-8 events being dropped it only drops 0-3. 

```
Agent.check = function() {
  var digest = [];
  //grab stored events from credentials
  var storedEvents = JSON.parse(this.credential('storedEvents')) || [];
  //iterate and add into digest array
  for(var i = storedEvents.length -1; i != -1; i--){
    digest.push(JSON.parse(storedEvents[i]));
  }
  //check if empty
  if (digest.length != 0){
	//Clear events out of credentials and send digest 
    this.credential('storedEvents', JSON.stringify(''));
    this.createEvent({'events': digest});

    }
    
};

Agent.receive = function() {
  var events = this.incomingEvents();
  //Turn payload into string and use hash to generate unique workerID
  var payload = JSON.stringify(events[0].payload);
  var hash = payload.hashCode();
  var workerID = payload.hashCode();
  //Check for memory lock/sleep if memory is locked
  while (this.memory('memLock') != workerID){
    sleepFor(Math.floor((Math.random() * 20) + (Math.random() * 5)));
	//set memory lock as worker ID
    if (this.memory('memLock') == 'false' || typeof this.memory('memLock') == 'undefined'){
        this.memory('memLock', workerID);
    }
  }
  //Grab currently stored events from credentials into an array
  var storedEvents = JSON.parse(this.credential('storedEvents')) || [];
  storedEvents.push(payload);
  //Encode updated array back into credentials
  this.credential('storedEvents', JSON.stringify(storedEvents));
  //remove lock
  this.memory('memLock', 'false');
  
}
//Some sleep function I copied off of google
function sleepFor( sleepDuration ){
    var now = new Date().getTime();
    while(new Date().getTime() < now + sleepDuration){ /* do nothing */ } 
}
//Some hash(hashes based on string) function I copied off of google 
String.prototype.hashCode = function(){
	var hash = 0;
	if (this.length == 0) return hash;
	for (i = 0; i < this.length; i++) {
		char = this.charCodeAt(i);
		hash = ((hash<<5)-hash)+char;
		hash = hash & hash; // Convert to 32bit integer
	}
	return hash;
}
```
If anyone could give me pointers to improve it or, give me a better solution altogether, I'd greatly appreciate it.

**Comment by dsander** [comment]
> It would seem that the problem is agents would copy the entire memory-update it- and set it as the new memory while other agents are currently doing the same thing.

Yes I think so too. We don't have a way to atomically update the update the database memory, so it has to be loaded into ruby, updated and written back.

> My reasoning is that if certain memory heavy agents are locked to a single worker, agents like Deduplication would be increasingly slow under high load. I run a deduplication agent that handles around 100 events per propagation, so that would take minutes as opposed to seconds to handle.

Especially when running multiple different DigestAgents that solution would get slower and slower. When just one Agent is used I don't think there would be much difference between ensuring only one `receive` function of the Agent can run at the same time or having that Agent only run in one worker.

> Also, I think i found another bug in the process. Originally, I was going to set the memory lock in credentials but it seems like agents cache credential key-value pairs in memory because I could not read an updated credential after its been read once before.

Right, that was designed as a feature, but I can see that it looks more like a bug in your use case.

> I wanted to create a memory 'stack' but I couldn't figure out a way to push an update into credentials(or memory) without copy-update-paste the entire thing.

Not sure if there is a way to do this with the databases we are supporting, redis certainly has data structures that would support it, but with MySQL and Postgres I think we are limited to a solution very similar to your JavaScriptAgent.

**Comment by Kristovich** [comment]
So here's an update. I am finally able to get all my events aggregated properly without dropping a single event properly without dealing with memory. This is definitely a roundabout solution though. So I have all my events piped into a liquid output agent which a website agents gets all the events into one key and finally a custom javascript agent puts it into an array. 
Here it is for anyone else that needs a solution:
Liquid output agent
`{{events| replace: '=>', ':'}}`
"last X events" dropdown
Website agent
```
"extract": {
    "events": {
      "regexp": ".*",
      "index": "0"
    }
  }
```
Javascript agent(applicable lines)
```
  var events = this.incomingEvents();
  var eventstring = events[0].payload["events"];
  var preArray = eventstring.split(',');
  var eventArray = JSON.parse(preArray)
```

I took a little look at Liquid output agent's code and it seems like it doesn't use memory at all in "last x events" mode. It seems to look at the agents its receiving from and get the last events emitted from them sorted by date and limited by count. Maybe Digest agent could do something similar? The only downside is that Liquid Output agent will keep its webpage populated as long as the agents it receives from has emitted events. This can be solved by limiting the "keep events" option in the source agent.

**Comment by dsander** [comment]
That is a clever solution!



> The only downside is that Liquid Output agent will keep its webpage populated as long as the agents it receives from has emitted events. This can be solved by limiting the "keep events" option in the source agent.

That limitation is also the reason why the LiquidOutputAgent does not need to store the received Events in it's memory. If it should output the received Events once it would need to keep track of the emitted events somehow.
